File: src/adventofcode2018/Day17.java
package adventofcode2018;

import java.awt.Point;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import utils.Input;

/**
 * https://adventofcode.com/2018/day/17
 */
public class Day17 {

    public static void main(String[] args) {
        List<String> input = new Input(2018, 17).strings();
        Pattern p = Pattern.compile("(.)=(\\d+), .=(\\d+)..(\\d+)");
        Set<Point> clay = new HashSet<>();
        for (String s : input) {
            Matcher m = p.matcher(s);
            m.find();
            int x = Integer.parseInt(m.group(2));
            int a = Integer.parseInt(m.group(3));
            int b = Integer.parseInt(m.group(4));
            for (int y=a; y<=b; y++) {
                clay.add(m.group(1).equals("x")
                        ? new Point(x, y)
                        : new Point(y, x));
            }
        }
        int minx = clay.stream().mapToInt(pt -> pt.x).min().getAsInt();
        int maxx = clay.stream().mapToInt(pt -> pt.x).max().getAsInt();
        int miny = clay.stream().mapToInt(pt -> pt.y).min().getAsInt();
        int maxy = clay.stream().mapToInt(pt -> pt.y).max().getAsInt();
        System.out.printf("x[%d:%d], y[%d:%d]", minx, maxx, miny, maxy);

        Set<Point> flow = new HashSet<>();
        Set<Point> water = new HashSet<>();
        flow.add(new Point(500, miny));
        Set<Point> wet = new HashSet<>(flow);
        for (;!flow.isEmpty();) {
            Set<Point> nextFlow = new HashSet<>();
            for (Point drop : flow) {
                if (drop.y > maxy) continue;
                wet.add(drop);
                if (!clay.contains(new Point(drop.x, drop.y+1)) && !water.contains(new Point(drop.x, drop.y+1))) {
                    nextFlow.add(new Point(drop.x, drop.y+1));
                    continue;
                }
                Integer borderLeft = drop.x - 1;
                for (;;) {
                    Point left = new Point(borderLeft, drop.y);
                    if (clay.contains(left)) {
                        break;
                    }
                    if (clay.contains(new Point(borderLeft, drop.y + 1))
                            || water.contains(new Point(borderLeft, drop.y + 1))) {
                        wet.add(left);
                        borderLeft--;
                    } else {
                        nextFlow.add(left);
                        borderLeft = null;
                        break;
                    }
                }
                Integer borderRight = drop.x + 1;
                for (;;) {
                    Point right = new Point(borderRight, drop.y);
                    if (clay.contains(right)) {
                        break;
                    }
                    if (clay.contains(new Point(borderRight, drop.y + 1))
                            || water.contains(new Point(borderRight, drop.y + 1))) {
                        wet.add(right);
                        borderRight++;
                    } else {
                        nextFlow.add(right);
                        borderRight = null;
                        break;
                    }
                }
                if (borderLeft != null && borderRight != null) {
                    for (int x = borderLeft+1; x <= borderRight-1; x++) {
                        water.add(new Point(x, drop.y));
                    }
                    nextFlow.add(new Point(drop.x, drop.y-1));
                }
            }
            flow = nextFlow;
        }
        System.err.println(wet.size());
        System.err.println(water.size());
    }
}

File: src/Python/2018/Day17.py

from utils import Input, printResult
import re

# https://adventofcode.com/2018/day/17

clay = set()
for m in Input(2018, 17).match("(.)=(\\d+), .=(\\d+)..(\\d+)"):
    x, a, b = int(m[2]), int(m[3]), int(m[4])
    for y in range(a, b+1):
        clay.add((x, y) if m[1] == "x" else (y, x))

miny = min(p[1] for p in clay)
maxy = max(p[1] for p in clay)

flow = set([(500, miny)])
water = set()
wet = set(flow)
while len(flow) > 0:
    nextFlow = set()
    for drop in flow:
        wet.add(drop)
        p = (drop[0], drop[1] + 1)
        if p not in clay and p not in water:
            if p[1] <= maxy:
                nextFlow.add(p)
            continue

        left = (drop[0] - 1, drop[1])
        while left not in clay:
            p = (left[0], drop[1] + 1)
            if p in clay or p in water:
                wet.add(left)
                left = (left[0] - 1, left[1])
            else:
                nextFlow.add(left)
                left = None
                break

        right = (drop[0] + 1, drop[1])
        while right not in clay:
            p = (right[0], drop[1] + 1)
            if p in clay or p in water:
                wet.add(right)
                right = (right[0] + 1, right[1])
            else:
                nextFlow.add(right)
                right = None
                break

        if left and right:
            water.update((x, drop[1]) for x in range(left[0]+1, right[0]))
            nextFlow.add((drop[0], drop[1]-1))

    flow = nextFlow

printResult(1, len(wet))
printResult(2, len(water))
